### PR TITLE
AP_Notify: Support for UAVCAN RGB LEDs

### PR DESCRIPTION
**Pull request by Programmer74**
Hello all,

This PR adds support for new notification device, UAVCAN RGB LED. This utilizes broadcasting ```1081.LightsCommand.uavcan``` message with a single ```SingleLightCommand```. Then UAVCAN LED-powered device can hear the message and update its LED to the status that ArduPilot has.
The message is sent each 50 ms.

The reason for this is that this will allow Edge GNSS to provide the user with the same ArduPilot status information as the Edge's default onboard LED.

Please note that internally the ```SingleLightCommand``` features ```RGB565``` color, not ```RGB888``` color. All the necessary conversions are implemented.

This was tested on Edge with custom firmware on Edge GNSS and everything was fine.

Best,
Nikita Tomilov

### CODE REVIEW

**Comment by magicrub** [comment]
@Programmer74  Functionally this looks OK but needs some minor changes.
1) The 5 commits should be squashed to 2.
2) the RGB888 -> RGB565 conversion is happening constantly in do_cyclic(). Makes sense to change the _notify_conf struct variables to store it in RGB565 format (and say that in the struct definition). Then do the conversion once in led_write(), or even better in hw_set_rgb().

I love reading your Pull Requests because they're put together so well!

**Comment by Programmer74** [comment]
@magicrub 
I agree with the second one, yes, I'll move the conversion.
As for the first one, I intentionally left it this way. But if necessary, I'll squash them to two, but to which two?
Like, the first commit adding a new AP_Notify device, and the second one utilizing the logic in AP_UAVCAN?

**Comment by magicrub** [comment]
@Programmer74 yes, two commitsL one for each directory

**Comment by olliw42** [comment]
I think this could/should be seriously thought through before putting such low level messages to life on the quick.

TransferPriority::OneLowerThanHighest is not appropriate, IMHO, it rather should have a very low priority, OneHigherThanLowest.

more relevant, I think one could think a bit into the future and have a clarification of the light_id in the ArduPilot ecosystems of lights ... there might be many other leds, e.g. on the motor arms, etc. pp.. That is, probably some sort of enum extending the ID constants of the standard set could/should be drafted. Or similar "lists".

Lastly, since this is a led which reflects a status I think one at least should evaluate if it wouldn't be more meaningful to emit the status on the CAN bus instead of the low-level result such as an rgb. That is, one could wonder if it couldn't be more profitable to draft a sort of notifier (and events) message which emits notifications (and events) on the bus, and it is then up to the node to translate that into proper indications. To me, the advantages of such a message appear obvious and compelling.

just my 2 cents

**Comment by magicrub** [comment]
Sending a heart beat of sorts is a good idea, with information like being Armed out not which any can define could listen to and independently act on.. such as motor arm lights

**Comment by Programmer74** [comment]
@olliw42
I agree with the priority and I'll probably add the enum later on.
I'm currently working on how to emit multiple led values instead of just one.

However, I don't think we should emit the ArduPilot status here instead of RGB values. Here we add 
```AP_UAVCAN: RGB LED``` as a device on UAVCAN bus and a ```AP_Notify: UAVCAN RGB LED``` to utilize that UAVCAN device, not the ```AP_UAVCAN: ArduPilot status``` device.
This is a nice idea, but I believe that this should be the subject for another PR.

**Comment by Programmer74** [comment]
@magicrub I moved the conversion to ```led_write()``` since this would be easier for future to upgrade if we will implement true ```RGB888``` message later on; therefore we wouldn't have any need to change the ```AP_Notify: UAVCAN_RGB_LED```.

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 1]
```diff
@@ -1035,4 +1081,37 @@ void AP_UAVCAN::update_mag_state(uint8_t node)
     }
 }
 
+bool AP_UAVCAN::led_write(uint8_t led_index, uint8_t red, uint8_t green, uint8_t blue) {
+    uavcan::equipment::indication::RGB565 color;
+    color.red = (red >> 3);
+    color.green = (green >> 2);
+    color.blue = (blue >> 3);
+    
+    if (!led_out_sem_take()) return false;
```
Comment (magicrub) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: This whole function looks or if order. The take should be first, then the array bounds check, then the shift. 
 Seems like the enable check is not even needed here

Comment (magicrub) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Ok, I understand the enable check. That part is fine. Works be nice to have a comment saying the send happens above and it only computers in for registering new devices

File: libraries/AP_UAVCAN/AP_UAVCAN.h [hunk 2]
```diff
@@ -41,6 +42,9 @@
 #define AP_UAVCAN_HW_VERS_MAJOR 1
 #define AP_UAVCAN_HW_VERS_MINOR 0
 
+#define AP_UAVCAN_MAX_LED_DEVICES 4
+#define AP_UAVCAN_LED_DELAY_MILLISECONDS 50
```
Comment (magicrub) [inline libraries/AP_UAVCAN/AP_UAVCAN.h]: 20Hz seems a little fast? Is that normal?

Comment (Programmer74) [reply libraries/AP_UAVCAN/AP_UAVCAN.h]: Yes, this was totally normal for my setup

**Comment by magicrub** [comment]
@EShamaev ping

**Comment by magicrub** [comment]
@EShamaev ping

**Comment by magicrub** [comment]
My PR https://github.com/ArduPilot/ardupilot/pull/7678 will cause a merge conflict on this PR so I made [this branch](https://github.com/magicrub/ardupilot/tree/pr-can-rgbled2) which is this branch rebased on top of it.

**Comment by magicrub** [comment]
Closing because this has a merge conflict that I fixed with PR https://github.com/ArduPilot/ardupilot/pull/7680

**Pull request closed without merging**